src/Gui/TaskView/TaskImageScale.cpp

```diff
@@ -85,4 +112,196 @@ void TaskImageScale::changeHeight()
     }
 }
 
+View3DInventorViewer* TaskImageScale::getViewer() const
+{
+    if (!feature.expired()) {
+        auto vp = Application::Instance->getViewProvider(feature.get());
+        auto doc = static_cast<ViewProviderDocumentObject*>(vp)->getDocument();
+        auto view = dynamic_cast<View3DInventor*>(doc->getViewOfViewProvider(vp));
+        if (view) {
+            return view->getViewer();
+        }
+    }
+
+    return nullptr;
+}
+
+void TaskImageScale::selectedPoints(size_t num)
+{
+    if (num == 1) {
+        ui->labelInstruction->setText(tr("Select second point"));
+    }
+    else if (num == 2) {
+        ui->labelInstruction->setText(tr("Enter distance value"));
+        ui->pushButtonScale->setEnabled(true);
+        ui->pushButtonScale->setText(tr("Accept"));
+        ui->pushButtonCancel->show();
+        ui->quantitySpinBox->setEnabled(true);
+        ui->quantitySpinBox->setValue(scale->getDistance());
+    }
+}
+
+void TaskImageScale::scaleImage(double factor)
+{
+    if (!feature.expired()) {
+        feature->XSize.setValue(feature->XSize.getValue() * factor);
+        feature->YSize.setValue(feature->YSize.getValue() * factor);
+
+        QSignalBlocker blockW(ui->spinBoxWidth);
+        ui->spinBoxWidth->setValue(feature->getXSizeInPixel());
+        QSignalBlocker blockH(ui->spinBoxHeight);
+        ui->spinBoxHeight->setValue(feature->getYSizeInPixel());
+    }
+}
+
+void TaskImageScale::startScale()
+{
+    scale->activate();
+    ui->labelInstruction->setText(tr("Select two points in the 3d view"));
```

Comment:
It took me a while to figure out how this was supposed to work. I think it might help to be more explicit about the tool's "thought process" -- maybe something like:
```suggestion
    ui->labelInstruction->setText(tr("Select two points on the image in the 3d view"));
```

Reply:
Maybe simpler and cleaner: `Select two points on the image`

Reply:
But conceptually what we expect a user to be doing is clicking two points on the image whose physical separation distance they know: the fact that the implementation doesn't care if they click on the image or not probably doesn't matter to someone using the tool. I can't imagine a scenario where someone would be clicking outside their image when using this tool.

Reply:
@wwmayer I don't disagree. But when we are end users, we are looking for simple prompts. Freecad very often has some complicated vocabulary that does not help in the end-user land.

Reply:
> But conceptually what we expect a user to be doing is clicking two points on the image whose physical separation distance they know:

The function just does what the old function did. Since I haven't heard of any complaints about this behaviour I don't know whether it's wise to suddenly change it.

The second point is that getting points from the image is way more expensive than getting it from the focal plane because you have to traverse the scene graph for every tiny mouse movement and depending on how many nodes there are the application may be dead slow. If you have ever worked with Coin's SoExtSelection node you will know what I mean.

Reply:
OK. If you say it isn't reasonable I believe you 👍 -- I have not worked with SoExtSelection. If it makes sense to everyone else then I will adapt.

Reply:
What could be done is adding an option where the user can decide if the line must completely lie on the plane. And in order to avoid the expensive traversal of the scene graph it could still compute the intersection point with the focal plane for mouse move events and for click events it does then compute the intersection with the image plane.

Reply:
There is no an option to let the user decide. By default it requires the points to be inside the image.

Reply:
I use this feature a lot. I never thought or wanted to use it extrapolating the plane or the area of the image. I understand we don't need to do those things because the physical dimension the user is going to enter should be something known about something inside the image. If the user selects a point out of the image, the dimension they know is not meaningful anymore and then this tool loses its value.

Reply:
How do you know that this is the only valid use case? What if a user wants an image to fit into a face of another object? In this case the user will click on the corner points of the face and they don't have to be necessarily inside the image plane.